### DISCUSSION TITLE
different query key for queries and mutations

### DISCUSSION THREAD

**Discussion by ejarquin-applaudo** [TanStack/query#2521] [General]
I wan to reuse the query key to get data from the server and modify that query after mutation, but I am seeing different query keys to the same query

![image](https://user-images.githubusercontent.com/74374394/127725450-40bba8b3-9244-4894-b1e0-ae497d4d4a3a.png)

Query key on the left panel:
`["/speedups/appdetails_appManager",{"customer_id":"55-6718-A000","frequency":30,"level":"appManager"}]
`

Query key on the right panel

```javascript
[
  "/speedups/appdetails",
  {
    "level": "appManager",
    "frequency": 30,
    "customer_id": "55-6718-A000"
  }
]
```

the first item of the array is   `"/speedups/appdetails",` but on the left panel is `/speedups/appdetails_appManager`

**Problem 1 with getQueryData**
Now, when I try to get cached data using` queryClient.getQueryData ` using the query key on the left panel, it doesnt work, I need to use the query key from the right panel.

**It doesnt work:**
```javascript
getQueryData([
  "/speedups/appdetails_appManager",
  {
    "level": "appManager",
    "frequency": 30,
    "customer_id": "55-6718-A000"
  }
])
```
**It works:**
```javascript
getQueryData([
  "/speedups/appdetails",
  {
    "level": "appManager",
    "frequency": 30,
    "customer_id": "55-6718-A000"
  }
])
```

**Problem 2 with setQueryData**
Now, the opposite happens here, when I try to update cached data using` queryClient.setQueryData ` using the query key on the right panel, it doesnt work, I need to use the query key from the left panel.

**It works:**
```javascript
setQueryData([
  "/speedups/appdetails_appManager",
  {
    "level": "appManager",
    "frequency": 30,
    "customer_id": "55-6718-A000"
  }
])
```
**It doesnt work:**
```javascript
setQueryData([
  "/speedups/appdetails",
  {
    "level": "appManager",
    "frequency": 30,
    "customer_id": "55-6718-A000"
  }
])
```
NOTE:
I am using the property: **queryKeyHashFn** to add the text appManager and modify the query key but Im seeing this behavior.
```javascript
  return useQuery([
  "/speedups/appdetails",
  {
    "level": "appManager",
    "frequency": 30,
    "customer_id": "55-6718-A000"
  }
], {
// here I have a function to get the new query key, I just append the `_appManager` text to the first item of the query key array
    queryKeyHashFn: getQueryKeyHash('appManager'),
    enabled: Boolean(params?.customerId),
    ...config,
  });
```
```javascript
export function getQueryKeyHash(submodule = '') {
  return (queryKey) => {
    const path = queryKey[0];
    const params = queryKey[1];
    return hashQueryKey([`${path}_${submodule}`, params]);
  };
}
```

**Comment by TkDodo** [comment]
> I am using the property: queryKeyHashFn to add the text appManager and modify the query key but Im seeing this behavior.

why are you doing that? I don't think that's what this function is for. It is usually just for making sure that the queryKey is serializable. You can't expect to be able to access elements with the _result_ of the hashQueryKey function. the result is always a string, but the input can be an array with objects as well.

so the query key you pass to getQueryData / setQueryData should always be the queryKey _before_ being processed by the queryKeyHashFn.

**Reply by ejarquin-applaudo** [reply]
mmm good point, I didn't know about that last point, thanks @TkDodo

**Comment by JuanmaMenendez** [comment]
hi, @TkDodo can a **mutation** and **query** have the same **key** without issue?

**Reply by TkDodo** [reply]
sure
